x/wasm/handler.go

```diff
@@ -2,215 +2,70 @@ package wasm
+			res, err = msgServer.InstantiateContract(sdk.WrapSDKContext(ctx), msg)
 		case *MsgExecuteContract:
-			return handleExecute(ctx, k, msg)
+			res, err = msgServer.ExecuteContract(sdk.WrapSDKContext(ctx), msg)
 		case *MsgMigrateContract:
-			return handleMigration(ctx, k, msg)
+			res, err = msgServer.MigrateContract(sdk.WrapSDKContext(ctx), msg)
 		case *MsgUpdateAdmin:
-			return handleUpdateContractAdmin(ctx, k, msg)
+			res, err = msgServer.UpdateAdmin(sdk.WrapSDKContext(ctx), msg)
 		case *MsgClearAdmin:
-			return handleClearContractAdmin(ctx, k, msg)
+			res, err = msgServer.ClearAdmin(sdk.WrapSDKContext(ctx), msg)
 		default:
 			errMsg := fmt.Sprintf("unrecognized wasm message type: %T", msg)
 			return nil, sdkerrors.Wrap(sdkerrors.ErrUnknownRequest, errMsg)
 		}
+
+		ctx = ctx.WithEventManager(filterMessageEvents(ctx))
+		return sdk.WrapServiceResult(ctx, res, err)
 	}
 }
 
-// filterMessageEvents returns the same events with all of type == EventTypeMessage removed.
+// filterMessageEvents returns the same events with all of type == EventTypeMessage removed except
+// for wasm message types.
 // this is so only our top-level message event comes through
-func filteredMessageEvents(manager *sdk.EventManager) []abci.Event {
-	events := manager.ABCIEvents()
-	res := make([]abci.Event, 0, len(events))
-	for _, e := range events {
-		if e.Type != sdk.EventTypeMessage {
-			res = append(res, e)
+func filterMessageEvents(ctx sdk.Context) *sdk.EventManager {
+	m := sdk.NewEventManager()
+	for _, e := range ctx.EventManager().Events() {
+		if e.Type == sdk.EventTypeMessage &&
+			!hasWasmModuleAttribute(e.Attributes) {
+			continue
 		}
+		m.EmitEvent(e)
 	}
-	return res
-}
-
-func handleStoreCode(ctx sdk.Context, k *Keeper, msg *MsgStoreCode) (*sdk.Result, error) {
-	senderAddr, err := sdk.AccAddressFromBech32(msg.Sender)
-	if err != nil {
-		return nil, sdkerrors.Wrap(err, "sender")
-	}
-	codeID, err := k.Create(ctx, senderAddr, msg.WASMByteCode, msg.Source, msg.Builder, msg.InstantiatePermission)
-	if err != nil {
-		return nil, err
-	}
-
-	ctx.EventManager().EmitEvents(sdk.Events{
-		sdk.NewEvent(
-			sdk.EventTypeMessage,
-			sdk.NewAttribute(sdk.AttributeKeyModule, ModuleName),
-			sdk.NewAttribute(types.AttributeKeySigner, msg.Sender),
-			sdk.NewAttribute(types.AttributeKeyCodeID, fmt.Sprintf("%d", codeID)),
-		),
-	})
-
-	return &sdk.Result{
-		Data:   []byte(fmt.Sprintf("%d", codeID)),
-		Events: ctx.EventManager().ABCIEvents(),
-	}, nil
+	return m
 }
 
-func handleInstantiate(ctx sdk.Context, k *Keeper, msg *MsgInstantiateContract) (*sdk.Result, error) {
-	senderAddr, err := sdk.AccAddressFromBech32(msg.Sender)
-	if err != nil {
-		return nil, sdkerrors.Wrap(err, "sender")
-	}
-	var adminAddr sdk.AccAddress
-	if msg.Admin != "" {
-		if adminAddr, err = sdk.AccAddressFromBech32(msg.Admin); err != nil {
-			return nil, sdkerrors.Wrap(err, "admin")
+func hasWasmModuleAttribute(attrs []abci.EventAttribute) bool {
```

Comment:
Why do we need this? I am a bit unsure and need to investigate the event/attribute system a bit more.

There were multiple sub-messages merging into each other before causing confusion. I will try some tests in a branch

Reply:
Saved follow up for #368 but not for 0.14.0 release